[Dang Hoang Duy] iP

DiligentPenguinn opened the pull request:
# Diligent Penguin
> If I feel unhappy, I do mathematics to become happy. If I am happy, I do mathematics to keep happy.
Quoted in P Turán, "The Work of Alfréd Rényi" [Source](https://mathshistory.st-andrews.ac.uk/Biographies/Renyi/quotations/#:~:text=If%20I%20feel%20unhappy%2C%20I,for%20turning%20coffee%20into%20theorems.) 

In this PR, I implement the following ~~modifications~~ enhancements:
- Implement **multiple new features** for the chatbot
- Add _some test cases_

Some notable methods:
1. Add a task (todo, deadline event)
2. Mark task as completed/uncompleted
3. Save/Load the list of tasks
4. Delete tasks

Here is a random code block (method `save` in `Storage` class)
```java
public void save(TaskList tasks) throws ChatBotException {
        try {
            FileWriter fw = new FileWriter(this.filePath);
            ArrayList<Task> arrayTask = tasks.getAllTasks();
            for (Task task: arrayTask) {
                fw.write(task.toSavedString() + "\n");
            }
            fw.close();
        } catch (IOException e) {
            throw new ChatBotException("Oops! Something went wrong, I cannot save the tasks list! "
                    + e.getMessage());
        }
}
```

New features:
- [x] Methods for managing tasks, events, and deadlines 🗒️
- [ ] Reminders ⏰(coming soon)

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

DiligentPenguinn committed:
Change chatbot's name and add a greet method

DiligentPenguinn committed:
add exit method

DiligentPenguinn committed:
add method echo, and implement necessary logic

DiligentPenguinn committed:
change chatbot's personality

DiligentPenguinn committed:
Create new class MyList to hold items

DiligentPenguinn committed:
add methos store and list, make use of MyList to store and print items

DiligentPenguinn committed:
Added new class Task to manage marking, and change TaskList to use Task

DiligentPenguinn committed:
Add mark method

DiligentPenguinn committed:
implement logic to treat mark commands

DiligentPenguinn committed:
implement logic to treat unmark commands

DiligentPenguinn committed:
Modify TaskList to accept tasks of different types

DiligentPenguinn committed:
change the logic for adding task

DiligentPenguinn committed:
return total tasks after storing

DiligentPenguinn committed:
Implement testing

DiligentPenguinn committed:
Create custom exception, handle some simple exceptions

DiligentPenguinn committed:
Update test case

DiligentPenguinn committed:
Add delete functionality

DiligentPenguinn committed:
Update test case

DiligentPenguinn committed:
Use enum for task types

DiligentPenguinn committed:
DiligentPenguin class: change all methods to non-static

DiligentPenguinn committed:
Event.java: fix a typo in toString method

DiligentPenguinn committed:
Add method to save tasks to a file, with helper methods

DiligentPenguinn committed:
Add functionality to load existing tasks data file and logic to handle exceptions

DiligentPenguinn committed:
Merge branch 'branch-Level-7'

DiligentPenguinn committed:
Add functionality to recognize date

DiligentPenguinn committed:
Merge branch 'branch-Level-8'

DiligentPenguinn committed:
Move save method to Storage.java

DiligentPenguinn committed:
Move load functionality  to Storage.java

DiligentPenguinn committed:
Move file create method to Storage.java

DiligentPenguinn committed:
Extract logic to Ui.java

DiligentPenguinn committed:
Extract command logic to Parser.java

DiligentPenguinn committed:
Remove unnescessary lines

DiligentPenguinn committed:
Merge branch 'branch-A-MoreOOP'

DiligentPenguinn committed:
Divide classes into packages

DiligentPenguinn committed:
Merge branch 'branch-A-Packages'

DiligentPenguinn committed:
Divide classes into packages 2

DiligentPenguinn committed:
Merge branch 'branch-A-Packages'

DiligentPenguinn committed:
Merge branch 'add-gradle-support' of https://github.com/DiligentPenguinn/ip

DiligentPenguinn committed:
Add test cases for Event.java and Deadline.java

DiligentPenguinn committed:
Change gradle build to generate Jar file

DiligentPenguinn committed:
Add JavaDoc comments to classes/methods in task package

DiligentPenguinn committed:
Add JavaDoc comments to remaining classes/methods

DiligentPenguinn committed:
Reformat code to follow coding standard

DiligentPenguinn committed:
Add functionality to find tasks by keywords

DiligentPenguinn committed:
Merge branch 'branch-A-JavaDoc'

DiligentPenguinn committed:
Resolve conflict when merging A-CodingStandard to master

DiligentPenguinn committed:
Resolve conflict when merging A-CodingStandard to master

DiligentPenguinn committed:
Resolve conflict when merging Level-9 to master

DiligentPenguinn committed:
Modify code based on checkstyle

DiligentPenguinn committed:
Implement GUI with JavaFX

DiligentPenguinn committed:
Add separate classes for different commands (currently working)

carineang commented on src/main/java/diligentpenguin/DiligentPenguin.java:
```diff
@@ -0,0 +1,168 @@
+    public String getResponse(String input) {
+        try {
+            return parser.parse(input);
+        } catch (Exception e) {
```
May consider adding specific exceptions such as ChatBotException or DateTimeParseException etc.

carineang commented on src/main/java/diligentpenguin/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,25 @@
+        this.deadline = deadline;
+    }
+    public static String getCommandInfo() {
+        return "This command adds a new Deadline task onto the list"
```
Consider checking this sentence, if it correspond to the logic in Parser.java:
```
if (!item.contains("/by")) {
    throw new ChatBotException(DeadlineCommand.getCommandInfo());
} 
```

carineang reviewed (commented):
Generally well done :)

nguyenvukhang commented on src/main/java/diligentpenguin/task/Task.java:
```diff
@@ -0,0 +1,78 @@
+        return this.isDone;
+    }
+
+    public static DateTimeFormatter getInputFormatter() {
```
Consider renaming to `getDateTimeFormatter` for clarity

DiligentPenguinn committed:
Delete unused method

DiligentPenguinn committed:
Add assertions

Some methods require preconditions to execute correctly.

Without explicit checks, these requirements may not be satisfied,
leading to unknown buggy behaviours.

Let's add assertions to ensure the preconditions are met before
executing the methods.

Using assertions is a common way to define assumptions so that
the runtime can verify them.

DiligentPenguinn committed:
Store magic numbers as constants

DiligentPenguinn committed:
Add different customized exception classes to ensure SLAP

DiligentPenguinn committed:
Merge pull request #1 from DiligentPenguinn/branch-A-Assertions

Add insertions

DiligentPenguinn committed:
Resolve conflict when merging remote master to branch A-CodeQuality

DiligentPenguinn committed:
Merge pull request #2 from DiligentPenguinn/branch-A-CodeQuality

Improve Code Quality

DiligentPenguinn committed:
Set up CI

DiligentPenguinn committed:
Add an extension: C-Update

DiligentPenguinn committed:
Fix style

DiligentPenguinn committed:
Fix checkstyle files

DiligentPenguinn committed:
Enhance code quality with SLAP, and prevent long methods

DiligentPenguinn committed:
Enhance code quality

DiligentPenguinn committed:
Fix style issues

DiligentPenguinn committed:
Add more exceptions handling for different commands

DiligentPenguinn committed:
Add more exception handling for datetime related errors

DiligentPenguinn committed:
Fix style

DiligentPenguinn committed:
Improve code quality with SLAP

DiligentPenguinn committed:
Add more JUnit Tests for most important classes

DiligentPenguinn committed:
Fix styles

DiligentPenguinn committed:
Enhance GUI

DiligentPenguinn committed:
Enhance GUI 2

DiligentPenguinn committed:
Add icon image

DiligentPenguinn committed:
Give credits to reused code

DiligentPenguinn committed:
Give credits to reused code 2

DiligentPenguinn committed:
Add user guide

DiligentPenguinn committed:
Remove unnescessary lines in user guide

DiligentPenguinn committed:
Fix coding standard

DiligentPenguinn committed:
Modify the user guide

DiligentPenguinn committed:
Enhance user guide

DiligentPenguinn committed:
Modify user guide

DiligentPenguinn committed:
Modify user guide

DiligentPenguinn committed:
Modify user guide

DiligentPenguinn committed:
Fix javafx version
